resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala

```diff
@@ -737,13 +735,75 @@ private[spark] class MesosClusterScheduler(
     if (index != -1) {
       pendingRetryDrivers.remove(index)
       pendingRetryDriversState.expunge(id)
+      suppressOrRevive()
       true
     } else {
       false
     }
   }
 
-  def getQueuedDriversSize: Int = queuedDrivers.size
-  def getLaunchedDriversSize: Int = launchedDrivers.size
-  def getPendingRetryDriversSize: Int = pendingRetryDrivers.size
+  private def copyBuffer(buffer: ArrayBuffer[MesosDriverDescription]):
+      ArrayBuffer[MesosDriverDescription] = {
+    val newBuffer = new ArrayBuffer[MesosDriverDescription](buffer.size)
+    buffer.copyToBuffer(newBuffer)
+    newBuffer
+  }
+
+  /**
+   * Check if the task state is a recoverable state that we can relaunch the task.
+   * Task state like TASK_ERROR are not relaunchable state since it wasn't able
+   * to be validated by Mesos.
+   */
+  private def isFailure(state: MesosTaskState): Boolean = {
+    state == MesosTaskState.TASK_FAILED ||
+      state == MesosTaskState.TASK_LOST
+  }
+
+  private def shouldSuppress: Boolean = {
+    return queuedDrivers.isEmpty && pendingRetryDrivers.isEmpty
+  }
+
+  private def suppressOrRevive(): Unit = {
```

Comment:
This is the meat of the functionality change.  We call this whenever the state of `queuedDrivers` or `pendingRetryDrivers` has changed.